File: scala/hamming/src/main/scala/hamming.scala
object Hamming {
  def compute(first: String, second: String, distance: Int = 0): Option[Int] = (first, second) match {
    case (a, b) if a == b => Some(distance)
    case (a, b) if a.length != b.length => None
    case (a, b) if a.head == b.head => compute(a.tail, second.tail, distance)
    case _ => compute(first.tail, second.tail, distance + 1)
  }
}

File: ocaml/hamming/hamming.ml
type nucleotide = A | C | G | T

let hamming_distance first second =
  if List.length first != List.length second then None
  else let rec calculate first second distance =
    match first, second with
      | [], _ -> distance
      | a::a_tail, b::b_tail when a == b -> calculate a_tail b_tail distance
      | _::a_tail, _::b_tail -> calculate a_tail b_tail (distance + 1)
      | _ -> assert false
    in
      let distance = calculate first second 0 in
      Some (distance)
